### PR TITLE
Add Pytorch Lightning spark estimator

### PR DESCRIPTION
**Pull request by irasit**
TorchEstimator(model) can be either a traditional PyTorch module or a PyTorch Lightning LightningModule.

When PyTorch Lightning is used, the optimizer and loss params must be omitted.

## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes # (issue).

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by EnricoMi** [comment]
You are running a lot of test here. To reduce turnaround time on your side and costs on our side, please consider reducing the test space to test images that are most affected by your changes, e.g. different Spark versions and GPU.

For this you can remove some lines from here until your code is complete:
https://github.com/horovod/horovod/blob/6f0bb9fae826167559501701d4a5a0380284b5f0/.buildkite/gen-pipeline.sh#L20-L59

And the run

    BUILDKITE_PIPELINE_SLUG=SLUG BUILDKITE_BRANCH=BRANCH .buildkite/gen-pipeline.sh > test/single/data/expected_buildkite_pipeline.yaml

Commit this and there will be fewer tests.

Remember to revert this commit later before merging into master, which will finally run all tests again.

File: .buildkite/gen-pipeline.sh
Comment (EnricoMi) [inline .buildkite/gen-pipeline.sh:366]: the test title should be unique, please add 'Lightning' it.

File: .travis.yml
Comment (EnricoMi) [inline .travis.yml:49]: you can append `pytorch_lighning` to the line above, as you do in `Dockerfile.cpu`

File: .travis.yml
Comment (EnricoMi) [inline .travis.yml:21]: Can you move `PYTORCH_LIGHTNING' one line up, so it aligns with `PYTORCH`?

File: docker-compose.test.yml
Comment (EnricoMi) [inline docker-compose.test.yml:16]: Can you move `PYTORCH_LIGHTNING` below `PYTORCH` here as well?

File: docker-compose.test.yml
Comment (EnricoMi) [inline docker-compose.test.yml:140]: Are there cuda versions for pytorch_lightning, or is this pure python on top of the respective pytorch package (which has the cuda bit)?

File: horovod/spark/common/params.py
Comment (EnricoMi) [inline horovod/spark/common/params.py:57]: why is the string converter removed?

File: horovod/spark/common/store.py
Comment (EnricoMi) [inline horovod/spark/common/store.py:90]: typo `chceckpoints`

File: horovod/spark/common/store.py
Comment (EnricoMi) [inline horovod/spark/common/store.py:239]: @tgaddair is this a breaking change?

File: test/single/test_ray.py
Comment (EnricoMi) [inline test/single/test_ray.py:31]: does this fix some specific transient ray test errors?

**Review by EnricoMi — CHANGES_REQUESTED** [review]

File: test/single/test_ray.py
Comment (irasit) [reply test/single/test_ray.py:31]: It fix the case in local. But the root cause of buildkite failure is the lightning version missmatch.

File: horovod/spark/common/store.py
Comment (irasit) [reply horovod/spark/common/store.py:239]: It was, I fixed it later.

File: horovod/spark/common/params.py
Comment (irasit) [reply horovod/spark/common/params.py:57]: validation: Optional validation column name (string) where every row in the column is either 1/True or 0/False,
                    or validation split (float) giving percent of data to be randomly selected for validation.

Comment (irasit) [reply horovod/spark/common/params.py:57]: It can be a string for col name or float for split.

File: docker-compose.test.yml
Comment (irasit) [reply docker-compose.test.yml:140]: This is pure python on top, it will relay on horovod backend and torch for GPU training.

File: .travis.yml
Comment (irasit) [reply .travis.yml:21]: sure

File: docker-compose.test.yml
Comment (irasit) [reply docker-compose.test.yml:16]: sure

File: .travis.yml
Comment (irasit) [reply .travis.yml:49]: sure

File: .buildkite/gen-pipeline.sh
Comment (irasit) [reply .buildkite/gen-pipeline.sh:366]: name changed.

File: test/single/test_ray.py
Comment (tgaddair) [reply test/single/test_ray.py:31]: I've also seen some random flakiness with ray in tests that requires changing these params. I think this is fine if it unblocks for now.

**Review by tgaddair — APPROVED** [review]
This is awesome!  Nice work @irasit and @chongxiaoc!

**Comment by tgaddair** [comment]
@EnricoMi does everything look good to you?

File: .buildkite/gen-pipeline.sh
Comment (EnricoMi) [inline .buildkite/gen-pipeline.sh:77]: Do build times really go up this much with your changes?

File: examples/pytorch/pytorch_lightning_mnist.py
Comment (EnricoMi) [inline examples/pytorch/pytorch_lightning_mnist.py:127]: this commented code should be removed

File: horovod/spark/common/params.py
Comment (EnricoMi) [reply horovod/spark/common/params.py:57]: and with that converter it always turned it into strings?

File: horovod/spark/common/store.py
Comment (EnricoMi) [reply horovod/spark/common/store.py:239]: @tgaddair?

File: Dockerfile.cpu
Comment (EnricoMi) [inline Dockerfile.cpu:6]: can you move this one line up?

File: Dockerfile.gpu
Comment (EnricoMi) [inline Dockerfile.gpu:7]: can you move this one line up?

File: Dockerfile.test.cpu
Comment (EnricoMi) [inline Dockerfile.test.cpu:15]: can you move this one line up?

File: Dockerfile.test.gpu
Comment (EnricoMi) [inline Dockerfile.test.gpu:16]: can you move this one line up?

**Review by EnricoMi — COMMENTED** [review]
some minor open questions and changes

File: .buildkite/gen-pipeline.sh
Comment (irasit) [reply .buildkite/gen-pipeline.sh:77]: Saw it used more time, but could be one time spike. Will roll it back and try

**Comment by irasit** [comment]
Fixed all the comment. Seems pytorch-lightning 1.2.8 dose not contain the hvd.join(local_rank) fix. Need to wait for the next version.

**Review by EnricoMi — APPROVED** [review]
Excellent, approved, When the release has happened, please upgrade to get the tests green.

Can you turn this into a draft for the time being, so we do not accidentally merge it into master and break tests?

**Comment by EnricoMi** [comment]
I have turned the PR into a draft until the release.

**Pull request merged**
